### PR TITLE
i#3068: API for accessing cache model metrics

### PR DESCRIPTION
**Pull request by mat-kalinowski**
This patch contains implementation of the API for accessing separate
metrics gathered during the simulation.

Functions for accessing metrics were added for all caches in the
default configuration: L1I, L1D and LL. Also additional functions
for getting configuration options of the cache_simulator_t.

Issue: #3068

### CODE REVIEW

File: clients/drcachesim/simulator/cache_simulator.h
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_simulator.h:68]: In our usage we routinely have a 3-level cache with a per-core L2 in between L1{I,D} and LLC (created with a config file).  Possibilities:

+ Add a get_l2_metric().  To take in a core index, it would have to only support direct parents of L1* and so would not support arbitrary cache layouts.
+ Export an iterator over all caches (split into l1, llc, and "other"==l2 separate iterators?) which returns some kind of key (the cache name?) to then pass back to query a metric.
+ Take in the cache by name (the name from the config file) -- less usable by itself without an iterator.

File: clients/drcachesim/simulator/cache_simulator.h
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_simulator.h:77]: Since cache_simulator_knobs_t is already public -- should the knobs be returned?  IIRC a config file setup does set the key knobs fields?

File: clients/drcachesim/simulator/caching_device_stats.h
Comment (derekbruening) [inline clients/drcachesim/simulator/caching_device_stats.h:147]: Please add some kind of test.  Maybe in drcachesim_unit_tests or a new similar test?  Xref #4842.

**Review by derekbruening — COMMENTED** [review]
Thank you for the contribution.  I have a few suggestions.

**Comment by AssadHashmi** [comment]
add to whitelist

**Comment by derekbruening** [comment]
Looks like #4954 was hit on A64 Jenkins -- maybe should be added to the known-flaky list.

**Comment by derekbruening** [comment]
run arm tests

File: clients/drcachesim/simulator/cache_simulator.h
Comment (mat-kalinowski) [reply clients/drcachesim/simulator/cache_simulator.h:68]: Thanks for the suggestions! Sorry for the delay, I had few days off. I like the flexibility of accessing stats in arbitrary cache model and also having really simple interface. I thought about having common function for all caches, something like `int_least64_t get_cache_metric(unsigned level, unsigned core, metric_t metric)`
Does separation for instruction and data caches makes sense only for L1 cache in any model ?

Comment (derekbruening) [reply clients/drcachesim/simulator/cache_simulator.h:68]: I think the current code only supports i vs d at L1 but if it's easy not to assume that here that might help future support.

> something like `int_least64_t get_cache_metric(unsigned level, unsigned core, metric_t metric)`

Would this support asymmetric layouts?  I guess it would if you always walk from the core up `level` levels.

File: clients/drcachesim/simulator/caching_device_stats.h
Comment (mat-kalinowski) [reply clients/drcachesim/simulator/caching_device_stats.h:147]: I have added unit test in the `drcachesim_unit_tests`. I couldn't test `Invalidates` metric properly because of the bug/issue that I described here: https://github.com/DynamoRIO/dynamorio/issues/5031

File: clients/drcachesim/simulator/cache_simulator.h
Comment (mat-kalinowski) [reply clients/drcachesim/simulator/cache_simulator.h]: Yes, knobs fields are set by config file setup. I have added suitable method.

File: clients/drcachesim/simulator/cache_simulator.h
Comment (mat-kalinowski) [reply clients/drcachesim/simulator/cache_simulator.h]: I tried to make this code as flexible as possible. Current approach should work with split cache of the higher level in simple models, when given higher level split cache is an ancestor of L1 (for example L1i -> L2i).

Comment (mat-kalinowski) [reply clients/drcachesim/simulator/cache_simulator.h]: Can you provide example of some simple asymmetric layout ? My current approach is to walk up from L1 cache and I think it should work properly.

Comment (derekbruening) [reply clients/drcachesim/simulator/cache_simulator.h]: > Can you provide example of some simple asymmetric layout ? My current approach is to walk up from L1 cache and I think it should work properly.

Never used or seen anything too weird with dr$sim, was just speculating about possible corner cases.  Such a walk should work with any layout though it seems.

**Comment by mat-kalinowski** [comment]
Right now I am handling errors like this: 
```
if (parent == NULL) {
    ERRMSG("Wrong cache level: %u.\n", core);
    return 0;
}
```
I thought also about returning -1 without printing error message. Which approach you think is better? Also do you have some comments about newest commit with changes ?

**Comment by derekbruening** [comment]
> Right now I am handling errors like this:
> 
> ```
> if (parent == NULL) {
>     ERRMSG("Wrong cache level: %u.\n", core);
>     return 0;
> }
> ```
> 
> I thought also about returning -1 without printing error message. Which approach you think is better? 

For get_cache_metric() I would think of it as a library routine likely to be used by third-party code and as such I'd return an error rather than printing an error.

> Also do you have some comments about newest commit with changes ?

Oh, was waiting for a (re-)review request (https://dynamorio.org/page_code_reviews.html#autotoc_md114 "When the requested changes have been pushed, you can ask for a re-review from the reviewers. This can be done by clicking the re-review button next to the reviewers' name on the right sidebar on the pull request page.")

File: clients/drcachesim/simulator/cache_simulator.h
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_simulator.h:46]: Safer to put inside the class and not in the global namespace (xref #4343).  Or see other suggestions for the enum below.

File: clients/drcachesim/simulator/cache_simulator.h
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_simulator.h:66]: I would re-order the parameters with the metric first and grouping the 3 identifiers of which cache (level, core, split) together.

File: clients/drcachesim/simulator/caching_device_stats.h
Comment (derekbruening) [inline clients/drcachesim/simulator/caching_device_stats.h:53]: I know the existing code has global scope enums but we're trying to avoid that going forward (xref #4343), and here we have very simple names that could conflict.  We do require C++11 so this could be a scoped enum (`typedef enum class metric_name_t`) where it would require the prefix `metric_name_t::HITS`; or at least use a prefix like the existing enums so `METRIC_HITS`; or something, to avoid global symbols with very simple names like `HITS` which could conflict (this code is statically linked into some very large code bases).

File: clients/drcachesim/simulator/cache_simulator.cpp
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_simulator.cpp:605]: Xref the top-level thread: I would vote for returning an error.  Could go so far as error enum types; or just return -1 (document that's the error return code) either w/ or w/o the print (could see the print useful if there's no errno; could see it annoying linked into a large system that maybe doesn't know the cache layout and is passing all possible levels and using return code to find the bounds of the layout -- there an errno would be nice).

Ditto below.

File: clients/drcachesim/simulator/caching_device_stats.h
Comment (derekbruening) [reply clients/drcachesim/simulator/caching_device_stats.h:147]: OK, best to put a `XXX i#5031: ` or `TODO i#5031: ` comment in the test about adding that metric.

**Review by derekbruening — COMMENTED** [review]
Thank you for adding the test.  Just a few comments mostly about names and then should be good to go.

**Comment by mat-kalinowski** [comment]
I've changed enum types to the `enum class` to prevent polluting global namespaces. I've implemented `stats_error_t` as a plain enum and added `STATS_` prefix because I needed to return it as an int from the `get_cache_metric` function.

File: clients/drcachesim/simulator/cache_simulator.h
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_simulator.h:78]: Please comment that a negative return value is an error and is of type stats_error_t

File: clients/drcachesim/simulator/cache_simulator.h
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_simulator.h:50]: Use the name `get_cache_metric` rather than a description so a reader knows precisely which one

**Review by derekbruening — APPROVED** [review]
I just have some comment suggestions, otherwise looks good.  Thank you for the contribution.

**Comment by mat-kalinowski** [comment]
run arm tests

**Comment by derekbruening** [comment]
Thanks again for the contribution.

One nit: For future merges please remove the duplicate title line (xref https://dynamorio.org/page_code_reviews.html#autotoc_md117):

```
commit 9574fca502ee03661dfd788f8744a8881f19eb63

    i#3068: API for accessing cache model metrics (#5020)
    
    i#3068: API for accessing cache model metrics
    
    This patch contains implementation of the API for accessing separate
    metrics gathered during the simulation.
    
    Functions for accessing metrics were added for all caches in the
    default configuration: L1I, L1D and LL. Also additional functions
    for getting configuration options of the cache_simulator_t.
    
    Issue: #3068
```

**Pull request merged**
